NEXT-12792  [NEW] Administration: Using utility functions (#149)
shopware/docs: guides/plugins/plugins/administration/using-utils.md

# Using Utilities

Utility functions in the Shopware 6 Administration are registered to the Shopware object and are therefore accessible everywhere in the Administration.
They provide many useful shortcuts for common tasks, see all of them [here](./../../../../resources/references/core-reference/administration-reference/utils.md) in our reference section.

## Prerequisites

All you need for this guide is a running Shopware 6 instance, the files and preferably a registered module. Of course you'll have to understand JavaScript. However, that's a prerequisite for Shopware as a whole and will not be taught as part of this documentation.

## Accessing the Utility functions

As noted in the introduction utils in the Shopware 6 Administration are bound to the Shopware global object. Learn more about that [here](./the-shopware-object.md).

> In this very short example we'll show you how to use a function of utils, in this case the `capitalizeString` function. The `capitalizeString` function in turn calls the [`lodash capitalize`](https://lodash.com/docs/4.17.15#capitalize) function wich as the name implies capitalizes strings.

Reviewer: ```suggestion
In this very short example we'll show you how to use a function of utility functions, in this case the `capitalizeString` function. The `capitalizeString` function in turn calls the [`lodash capitalize`](https://lodash.com/docs/4.17.15#capitalize) function wich as the name implies capitalizes strings.
```

Somehow, this suggestion is still nowhere to be found.

{% code title="<plugin-root>/src/Resources/app/administration/app/src/component/swag-basic-example/index.js" %}
```javascript
const { Component, Utils } = Shopware;

Component.register('swag-basic-example', {

    data: function () {
        return {
            string: 'hello',
            capitalizedString: undefined
        }
    },

    created: function () {
        this.capitalize()
    },

    methods: {
        capitalize() {
            this.capitalizedString = Utils.string.capitalizeString(this.string);
        }
    }
});
```
{% endcode %}

## An overview of all the functions

In the example before we just used the `deepCopy` function, but there more functions than reasonable to display in this guide.

> If want to look up at all available utility functions, take a look at [the code that registers all of them](https://github.com/shopware/platform/blob/v6.3.4.1/src/Administration/Resources/app/administration/src/core/service/util.service.js).

Reviewer: Please reference yours here

> # Using Utils

Reviewer: ```suggestion
# Using Utilities
```

> Utils in the Shopware 6 Administration are registered to the Shopware object and are therefore accessible everywhere in the Administration.

Reviewer: ```suggestion
Utility functions in the Shopware 6 Administration are registered to the Shopware object and are therefore accessible everywhere in the Administration.
```

## Prerequisites

All you need for this guide is a running Shopware 6 instance, the files and preferably a registered module. Of course you'll have to understand JavaScript. However, that's a prerequisite for Shopware as a whole and will not be taught as part of this documentation.

> ## Accessing the Utils

Reviewer: ```suggestion
## Accessing the Utility functions
```

As noted in the introduction utils in the Shopware 6 Administration are bound to the Shopware global object. Learn more about that [here](./the-shopware-object.md).

> In this very short example we'll show you how to use a function of utils, in this case the `deepCopy` function.

Reviewer 2: perhaps you should describe in a few words what `deepCopy` does

Reviewer: I agree, please explain your example in more details, in your own words. Or at least use comments to emphasize the important aspects of this file. I don't like the usage of a plain file as example, without any explanation on it.

Author: well i changened the example and wrote a sentence about it

{% code title="<plugin-root>/src/Resources/app/administration/app/src/component/swag-basic-example/index.js" %}
```javascript
const { Component, Utils } = Shopware;

Component.register('swag-basic-example', {

    data: function () {
        return {
            testObject: {
                testValue0: 0,
                testValue1: 1,
                testValue2: 2
            },

>             clonedObeject: {

Reviewer 2: ```suggestion
            clonedObject: {
```

            }
        }
    },

    created: function () {
        this.clone()
    },

    methods: {
        clone() {

>             this.clonedObeject = Utils.object.cloneDeep(this.testObject);

Reviewer 2: ```suggestion
            this.clonedObject = Utils.object.cloneDeep(this.testObject);
```

        }
    }
});
```
{% endcode %}

> ## An overview of all the functions

Reviewer: How many functions are there? I'd suggest using a table *in this article* to introduce all functions in a short manner, not just refering to a file. So the reader doesn't need to switch between articles and we don't get a direct dependency on the file in the core.

Author: about 50 or so and i think that is way to many to just, let alone explain

Reviewer: I agree but please consider my remark I wrote to you in Slack, after all.

Reviewer 3: I see why it is difficult to keep them listed here but I think you should at least list a few topics like hashing, type detection, currencies, file access, ... this will make it more likely to find this page wihtin a search.

Reviewer: How did we ended up here, now? @PaddyS were you able to help yet?

Author: i wrote the reference in the past 2 days

> In the example before we just used the `deepCopy` function, but there are a lot more.

Reviewer: ```suggestion
In the example before we just used the `deepCopy` function, but there more functions than reasonable to display in this guide.
```

> If want to have a look at the code that registers them all, you can look at it [here](https://github.com/shopware/platform/blob/trunk/src/Administration/Resources/app/administration/src/core/service/util.service.js).

Reviewer: ```suggestion
If want to have a look at the code that registers all available functions, you can look at it [here](https://github.com/shopware/platform/blob/trunk/src/Administration/Resources/app/administration/src/core/service/util.service.js).
```

> If want to have a look at the code that registers all available functions, you can look at it [here](https://github.com/shopware/platform/blob/trunk/src/Administration/Resources/app/administration/src/core/service/util.service.js).

Reviewer: ```suggestion
If want to look up at all available utility functions, take a look at [the code that registers all of them](https://github.com/shopware/platform/blob/trunk/src/Administration/Resources/app/administration/src/core/service/util.service.js).
```

## Accessing the Utils

As noted in the introduction utils in the Shopware 6 Administration are bound to the Shopware global object. Learn more about that [here](./the-shopware-object.md).

In this very short example we'll show you how to use a function of utils, in this case the `deepCopy` function.

{% code title="<plugin-root>/src/Resources/app/administration/app/src/component/swag-basic-example/index.js" %}
```javascript
const { Component, Utils } = Shopware;

Component.register('swag-basic-example', {

    data: function () {
        return {
            testObject: {
                testValue0: 0,
                testValue1: 1,
                testValue2: 2
            },
            clonedObeject: {

            }
        }
    },

    created: function () {
        this.clone()
    },

    methods: {
        clone() {
            this.clonedObeject = Utils.object.cloneDeep(this.testObject);
        }
    }
});
```
{% endcode %}

## An overview of all the functions

> In the example before we just used the `deepCopy` function, but there are are a lot more.
> To many of them to list them all here.

Reviewer 4: ```suggestion
In the example before we just used the `deepCopy` function, but there are a lot more.
```

> As noted in the introduction utils in the Shopware 6 Administration are bound to the Shopware global object. Learn more about that [here](./the-shopware-object.md).

Reviewer: ```suggestion
As noted in the introduction, utility functions in the Shopware 6 Administration are bound to the Shopware global object. Learn more about that [here](./the-shopware-object.md).
```

> In this very short example we'll show you how to use a function of utils, in this case the `capitalizeString` function. The `capitalizeString` function in turn calls the [`lodash capitalize`](https://lodash.com/docs/4.17.15#capitalize) function wich as the name implies capitalizes strings.

Reviewer: ```suggestion
In this very short example we'll show you how to use a function of utils, in this case the `capitalizeString` function. The `capitalizeString` function calls the [`lodash capitalize`](https://lodash.com/docs/4.17.15#capitalize) function which capitalizes strings, as the name implies.
```

## An overview of all the functions

In the example before we just used the `deepCopy` function, but there more functions than reasonable to display in this guide.

> If want to look up at all available utility functions, take a look at [the code that registers all of them](https://github.com/shopware/platform/blob/trunk/src/Administration/Resources/app/administration/src/core/service/util.service.js).

Reviewer: ```suggestion
If you want to look up at all available utility functions, take a look at [the code that registers all of them](https://github.com/shopware/platform/blob/trunk/src/Administration/Resources/app/administration/src/core/service/util.service.js).
```
